using_restricted_array.rb

```diff
@@ -38,34 +102,101 @@ def sort(array, length)
 # constant, adds an element with 'SPECIAL_VALUE' in the end. Assumes the array
 # to be sorted in ascending order.
 def delete(array, length, value_to_delete)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length do
+      if array[i] == value_to_delete
+        until array[i] == nil do
+          if array[i+1] == nil
+            array[i] = SPECIAL_VALUE
+          else
+            array[i] = array[i+1]
+          end
+          i += 1
+        end
+      else
+        i += 1
+      end
+  end
+  return array
+
+  # puts "NOT IMPLEMENTED"
 end
 
 # Restricted array cannot be resized. So, we workaround by having a convention
 # Convention: replace all values with 'SPECIAL_VALUE'
 # Empties the restricted array by making all values = SPECIAL_VALUE
 def empty(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length
+    array[i] = SPECIAL_VALUE
+    i += 1
+  end
+  # puts "NOT IMPLEMENTED"
 end
 
 # Finds and returns the largest value element in the array which is not 'SPECIAL_VALUE'
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  largest = array[0]
+  i = 1
+  while i < length
+    if array[i] > largest && array[i] != SPECIAL_VALUE
+      largest = array[i]
+    end
+    i += 1
+  end
+  return largest
+  # puts "NOT IMPLEMENTED"
 end
 
 # Insert value to insert at the correct index into the array assuming the array
 # is sorted in ascending manner.
 # Restricted arrays cannot be resized. Insert only if there is space in the array.
-# (Hint: if there are elements with 'SPECIAL_VALUE', there is no room to insert)
+# (Hint: if there are no elements with 'SPECIAL_VALUE', there is no room to insert)
 # All subsequent elements will need to be moved forward by one index.
 def insert_ascending(array, length, value_to_insert)
-  puts "NOT IMPLEMENTED"
+  print_array(array)
+  # puts "value to insert: #{value_to_insert}"
+  has_room = false
+  i = 0
+  until array[i] == nil || has_room == true do

```

Comment:
Another optimization suggestion: You could eliminate this loop (which is O(n) operations) since you have access to the length of the array. You could instead check on array[length-1] == SPECIAL_VALUE to set has_room to true. (This would be O(1) operation since length is provided.)

Reply:
Aha! I see what you're doing - going backwards to insert. Nice trick! Ignore my previous comment.